src/cleaning/cleaning.c

```diff
@@ -49,6 +49,156 @@ struct cleaning_policy_ops cleaning_policy_ops[ocf_cleaning_max] = {
 	},
 };
 
+void ocf_cleaning_setup(ocf_cache_t cache, ocf_cleaning_t policy)
+{
+	ENV_BUG_ON(policy >= ocf_cleaning_max);
+
+	if (!cleaning_policy_ops[policy].setup)
+		return;
+
+	cleaning_policy_ops[policy].setup(cache);
+}
+
+int ocf_cleaning_initialize(ocf_cache_t cache, ocf_cleaning_t policy,
+		int init_metadata)
+{
+	ENV_BUG_ON(policy >= ocf_cleaning_max);
+
+	if (!cleaning_policy_ops[policy].initialize)
+		return 0;
+
+	return cleaning_policy_ops[policy].initialize(cache, init_metadata);
+}
+
+void ocf_cleaning_deinitialize(ocf_cache_t cache, ocf_cleaning_t policy)
+{
+	ENV_BUG_ON(policy >= ocf_cleaning_max);
+
+	if (!cleaning_policy_ops[policy].deinitialize)
+		return;
+
+	cleaning_policy_ops[policy].deinitialize(cache);
+}
+
+int ocf_cleaning_add_core(ocf_cache_t cache, ocf_core_id_t core_id)
+{
+	ocf_cleaning_t policy = cache->conf_meta->cleaning_policy_type;
+
+	ENV_BUG_ON(policy >= ocf_cleaning_max);
+
+	if (!cleaning_policy_ops[policy].add_core)
+		return 0;
+
+	return cleaning_policy_ops[policy].add_core(cache, core_id);
+}
+
+void ocf_cleaning_remove_core(ocf_cache_t cache, ocf_core_id_t core_id)
+{
+	ocf_cleaning_t policy = cache->conf_meta->cleaning_policy_type;
+
+	ENV_BUG_ON(policy >= ocf_cleaning_max);
+
+	if (!cleaning_policy_ops[policy].remove_core)
+		return;
+
+	cleaning_policy_ops[policy].remove_core(cache, core_id);
+}
+
+void ocf_cleaning_init_cache_block(ocf_cache_t cache, ocf_cache_line_t cache_line)
+{
+	ocf_cleaning_t policy = cache->conf_meta->cleaning_policy_type;
+
+	ENV_BUG_ON(policy >= ocf_cleaning_max);
+
+	if (!cleaning_policy_ops[policy].init_cache_block)
+		return;
+
+	cleaning_policy_ops[policy].init_cache_block(cache, cache_line);
+}
+
+void ocf_cleaning_purge_cache_block(ocf_cache_t cache,
+		ocf_cache_line_t cache_line)
+{
+	ocf_cleaning_t policy = cache->conf_meta->cleaning_policy_type;
+
+	ENV_BUG_ON(policy >= ocf_cleaning_max);
+
+	if (!cleaning_policy_ops[policy].purge_cache_block)
+		return;
+
+	cleaning_policy_ops[policy].purge_cache_block(cache, cache_line);
+}
+
+void ocf_cleaning_purge_range(ocf_cache_t cache, ocf_core_id_t core_id,
+		uint64_t start_byte, uint64_t end_byte)
+{
+	ocf_cleaning_t policy = cache->conf_meta->cleaning_policy_type;
+
+	ENV_BUG_ON(policy >= ocf_cleaning_max);
+
+	if (!cleaning_policy_ops[policy].purge_range)
+		return;
+
+	cleaning_policy_ops[policy].purge_range(cache, core_id, start_byte,
+			end_byte);
+}
+
+void ocf_cleaning_set_hot_cache_line(ocf_cache_t cache,
+		ocf_cache_line_t cache_line)
+{
+	ocf_cleaning_t policy = cache->conf_meta->cleaning_policy_type;
+
+	ENV_BUG_ON(policy >= ocf_cleaning_max);
+
+	if (!cleaning_policy_ops[policy].set_hot_cache_line)
```

Comment:
Since this is a hot IO path, and we're not using `nop` by default I'd consider marking this expression as unlikely